Add `Target` and `Work` types

tcharding opened the pull request:
Ugh! 1600 lines of green and 1100 of red - my apologies.

Currently we use the `Uint256` type for proof-of-work calculations. It was observed in #1181 that providing a public 256 bit integer type like this implies that it is a general  purpose integer type. We do not want to provide a general purpose integer type (see the 1000 arithmetic functions on stdlib integer types for why not :)

Add two new opaque integer types `Target` and `Work`. These are the inverse of each other, both conceptually and mathematically.

There is a lot of code in this PR, sorry about that. At a high level the PR does:

- Add a `pow` module.
- Put a modified version of  `Uint256` in `pow`, making it private.
- Add two new wrapper types `Target` and `Work` that provide a very limited API specific to their use case. In particular there are methods on each to convert to the other.
- Only implement methods that we use on each type.

### Note

During development I got mixed up with the word "difficulty", I have discovered this has a very specific meaning in Bitcoin. Please see rustdocs on the `Target::difficulty` function for explanation of this term. For this reason we use the type `Work` defined as the inverse of target, and reserve "difficulty" for the Bitcoin concept.

Kixunil commented on src/blockdata/block.rs:
```diff
@@ -135,27 +82,21 @@ impl BlockHeader {
 
     /// Checks that the proof-of-work for the block is valid, returning the block hash.
-    pub fn validate_pow(&self, required_target: &Uint256) -> Result<BlockHash, util::Error> {
+    pub fn validate_pow(&self, required_target: &Target) -> Result<BlockHash, util::Error> {
```
Pass by value would feel much more natural here.

tcharding replied:
Just noticed this one, done as suggested.

Kixunil commented on src/blockdata/constants.rs:
```diff
@@ -56,8 +56,8 @@ pub const SUBSIDY_HALVING_INTERVAL: u32 = 210_000;
 /// In Bitcoind this is insanely described as ~((u256)0 >> 32)
-pub fn max_target(_: Network) -> Uint256 {
-    Uint256::from_u64(0xFFFF).unwrap() << 208
+pub fn max_target(_: Network) -> Target {
```
Why we have an unused parameter here?

tcharding replied:
Removed, thanks.

Kixunil commented on src/consensus/params.rs:
```diff
@@ -8,31 +8,31 @@
 
 /// Lowest possible difficulty for Mainnet. See comment on Params::pow_limit for more info.
-const MAX_BITS_BITCOIN: Uint256 = Uint256([
+const MAX_BITS_BITCOIN: Difficulty = Difficulty::from_array([
```
Sounds like this should be `Difficulty::MAINNET_MIN`, analogous below.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1164 @@
+
+macro_rules! impl_from_and_impl {
+    ($ty:ident) => {
+        impl From<[u64; 4]> for $ty {
```
Implicit endianess is not great, I'd prefer to remove it and have a methods similar to `from_be_bytes`, `to_be_bytes` and their `le` counterparts.

tcharding replied:
Added `from_be`, `from_le`, `to_be`, `to_le`.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1164 @@
+
+macro_rules! impl_core_traits {
+    ($ty:ident) => {
+        impl core::ops::Add<$ty> for $ty {
```
`Target` shouldn't implement `Add` because adding targets is non-sensical for reason I explained in the other PR. Same for sub. `Mul<Self>` and `Div<Self>` and `Rem<Self>` as well but `Mul<{integer}>` and `Div<{integer}>` are probably fine.

For `Difficulty`, adding and subtracting seems fine, other math operations with `Self` aren't. Multiplication/division by scalars should be OK.

tcharding replied:
Ah cool, thanks. I should have flagged this. I knew I didn't quite get which operations should be implemented for which types. Thanks.

tcharding replied:
I'm stuck with this (from `block.rs`)
```rust
    /// Computes the popular "difficulty" measure for mining.
    pub fn difficulty(&self) -> u64 {
        (Target::max() / self.target()).low_u64()
    }

    /// Returns the total work of the block.
    pub fn work(&self) -> Difficulty {
        self.target().to_difficulty()
    }
```
I'd like to move that logic onto the `Target` type so we don't have to implement `Div<Self>` for `Target` but I've already got
```rust
    /// Converts this [`Target`] to a [`Difficulty`].
    pub fn to_difficulty(&self) -> Difficulty {
        // 2**256 / (target + 1) == ~target / (target+1) + 1    (eqn shamelessly stolen from bitcoind)
        let mut ret = !self.0;
        let mut ret1 = self.0;
        ret1.increment();
        ret = ret / ret1;
        ret.increment();
        Difficulty(ret)
    }
```
I'm confused by the words difficulty and work.

tcharding replied:
I think the API of `Target` and `Difficulty` needs carefully checking by someone with more knowledge of Bitcoin than me. I'm really guessing with this PR.

Kixunil replied:
I always knew that roughly difficulty = 1/target but never the details. Got me thinking about it and I believe I figured it out:

If difficulty = 1/target then target 0 would imply infinite difficulty - impossible. But the comparison of the block hash against target uses `<=` operator which implies 0 target is finding preimage to sha256 which is not impossible, just maximally difficult. Therefore 1 is added making the range of results (0, 1].

We don't want to deal with floating point numbers in computers so we just multiply it by 2^256. Notably this overflows U256 when target is 0. The assumption is that we can never hit it because there will never be target 0. However I still think it may be a bug in the context of abstract thinking about Bitcoin.

Then there's the magic equation from Core:

* when numbers are represented in two's complement, you get the negative of a number by computing `!x + 1`
* therefore `!x = -x - 1`
* however we keep using unsigned integer, so we get multiplicative group of integers modulo `MAX + 1` arithmetic
* so !x == `(MAX + 1) + (-x - 1)`
* modifying right side of the equation: `!target / (target + 1) + 1 == (2^256 -target - 1) / (target + 1) + 1`
* `== (2^256 -target - 1) / (target + 1) + (target + 1)/(target + 1)`
* `==  (2^256 -target - 1 + target + 1) / (target + 1)`
* `== 2^256 /(target +1)` QED

However note that there's another overflow bug in division there `Target::MAX.to_difficulty()` would divide by 0.

What does it mean for the type system though? I think the unit of any intermediate step is meaningless and we should just internally work on raw `U256`. Therefore `to_difficulty` stays as you wrote above. The `difficulty` function looks wrong but maybe there's something that uses that representation? If yes we should probably have a `lossy_to_u64` method for that.

Regarding the overflow bugs, I guess division should be patched to return 0 on `Target::MAX` and `Difficulty` should contain `enum DifficultyInner { Value(U256), MaxPlusOne }`. And no, the overflow bugs are not a problem in Core because it only decodes compact target which can not represent `Target::Max` and has an explicit check for target 0 (returning 0).

tcharding replied:
Thanks man, this all took me a fair bit of thinking today :)

> The difficulty function looks wrong but maybe there's something that uses that representation? If yes we should probably have a lossy_to_u64 method for that.

I think this is aswered by https://bitcoin.stackexchange.com/questions/8806/what-is-difficulty-and-how-it-relates-to-target

I've updated the PR quite a bit to reflect my current understanding, please pick holes in it if you see them.

tcharding replied:
Currently we have:
```rust
    fn inverse(&self) -> Uint256 {
        // We should never have a target/work of zero so this doesn't matter that much but we define
        // the inverse of 0 as MAX.
        if *self == Uint256::ZERO {
            return Uint256::MAX;
        }

        // We define the inverse of max as 1.
        if *self == Uint256::MAX {
            return Uint256::ONE;
        }

        // And the inverse of 1 as max.
        if *self == Uint256::ONE {
            return Uint256::MAX;
        }

        let mut denom = *self;
        denom.increment();

        let mut ret = !*self / denom;
        ret.increment();

        ret
    }
```

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1164 @@
+            fn rem(self, other: $ty) -> $ty { $ty(self.0.rem(other.0)) }
+        }
+
+        impl core::ops::BitAnd<$ty> for $ty {
```
Bit ops kinda leak our representation, do we want to do that?

tcharding replied:
All deleted!

Kixunil commented on src/pow.rs:
````diff
@@ -0,0 +1,1164 @@
+    /// assert_eq!(x, Difficulty::from(2));
+    /// ```
+    #[inline]
+    pub fn increment(&mut self) { self.0.increment() }
````
This is surprising, should it be `pub`?

tcharding replied:
Removed.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1164 @@
+
+        // The mantissa is signed but may not be negative
+        if mant > 0x7FFFFF {
+            Target::default()
```
Shouldn't we return `Err` instead?

tcharding replied:
Based on this code comment I did not touch this function's code
```
        // This is a floating-point "compact" encoding originally used by
        // OpenSSL, which satoshi put into consensus code, so we're stuck
        // with it. The exponent needs to have 3 subtracted from it, hence
        // this goofy decoding code:
```

tcharding replied:
I found a proper answer,

> When parsing nBits, Bitcoin Core converts a negative target threshold into a target of zero, which the header hash can equal (in theory, at least).

ref: https://developer.bitcoin.org/reference/block_chain.html#target-nbits

Kixunil reviewed (commented):
Light review for now.

tcharding commented:
Thanks for your review @Kixunil, I've had another go at it. This is going to take a few rounds, sorry to be a burden. No rush on this one, please just review when it amuses you to do so.

tcharding commented:
Big re-write folks, please see updated PR description explaining my new found understanding :) Thanks for being patient with me while I learn Bitcoin at the same time as trying to design correct Bitcoin APIs :)

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+    };
+}
+
+/// Implements `ops::Mul` by integer type in both directions. Requires `u32::from($int)`.
```
Why it's restricted to `u32`? Would be nice to allow `u128`

tcharding replied:
Mentioned elsewhere but flagging here as well, I did all integer sizes upto `u64`. The reason being that I changed the internal representation to us `[u128; 2]` and we can multiply two `u64`s inside of a `u128`.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+            type Output = Self;
+
+            #[inline]
+            fn shl(self, shift: $int) -> Self { Self(self.0 << shift as usize) }
```
This is weird, shifts in `std` accept `u32` not `usize`

tcharding replied:
Yeah I thought as much but everytime I try to use `u32` I get in a knot of casting types, and removing casts, all over the place. I'll have another go now you've prompted me.

tcharding replied:
Fixed, all use `u32` now.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+
+impl_div!(Work, u8);
+impl_div!(Work, u16);
+impl_div!(Work, u32);
```
Looks like you could've saved duplication by putting mul and div into same macro and calling `impl_mul_div` from `do_impl`

tcharding replied:
At other iterations I' was supporting different sized integers for each, I'm still not totally sure I have all the correct (sane/meaningful) integers supported. Can merge them though if `u8` through `u32` is all we want for both.

tcharding replied:
We now have `impl_div_mul` for ints sized 8 to 32

tcharding replied:
And now also `u64`.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    pub fn difficulty(&self) -> u64 {
```
Should contain documentation about too low target causing overflow. But wouldn't it be better to return `None` if difficulty is too high? We could also return `u128` but that still overflows. Or perhaps saturating is better than potentially returning too low number.

tcharding replied:
I played around with this but have not come up with a good solution yet.

tcharding replied:
I've implemented saturation and added docs, not sure if this is the best solution.

Kixunil commented on src/pow.rs:
````diff
@@ -0,0 +1,1314 @@
+    /// let t = Target::from(0xdeadbeef);
+    /// assert_eq!(t.to_work().to_target(), t)
+    /// ```
+    pub fn to_work(&self) -> Work { Work(self.0.inverse()) }
````
By value seems more natural.

tcharding replied:
Done, thank you.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+        }
+
+        let mut denom = *self;
+        denom.increment();
```
Perhaps rename it to `wrapping_inc_assign()`?

tcharding replied:
Agree, done.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+        } else {
+            Ok(Self::_from_be_slice(bytes))
+        }
+    }
```
I'd prefer to remove `from_le_slice` and `from_be_slice` from everywhere. People can use `try_into()` if they have a slice argument and handle the error as neeeded.

tcharding replied:
Removed. While refactoring and unit testing I removed all usage of the `endian` module. I rekon we may be able to remove the module all together. Will investigate as a separate PR.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+        Uint256(slice)
+    }
+
+    fn _from_le_slice(bytes: &[u8]) -> Uint256 {
```
Accepting array would be nicer to the optimizer too.

tcharding replied:
This goes along with the idea that types of 32 bytes or less that implement `Copy` should be passed by value, right?

Kixunil replied:
Not specifically `Copy`, just various traits. 32 bytes is within limit.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+        slice
+            .iter_mut()
+            .rev()
+            .zip(bytes.chunks(8))
```
`chunks_exact`

tcharding replied:
Nice, done.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+    }
+
+    #[cfg(test)]
+    fn bit(&self, index: usize) -> bool {
```
Unimportant bikeshedding: `bit_at` seems better.

tcharding replied:
Leaving open for input from others.

apoelstra replied:
I also kinda prefer `bit_at`

tcharding replied:
Sweet, two for `bit_at`, that's good enough for me :)

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+    fn cmp(&self, rhs: &Uint256) -> cmp::Ordering {
+        // We need to manually implement ordering because we use little-endian
+        // and the auto derive is a lexicographic ordering(i.e. memcmp)
+        // which with numbers is equivalent to big-endian
```
I wonder if there is any advantage using little endian representation. There's at least the disadvantage that we can't derive `Ord`.

tcharding replied:
I'm pretty confident with all the new/improved unit tests that I can trivially switch it to big-endian, @apoelstra do you remember why we used little-endian originally?

apoelstra replied:
Probably because it was easier for me to implement some algorithms with little endian, or maybe just easier to think about stuff .... I'm pretty sure it's fine to switch to BE internally.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+
+/// Invalid slice length.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Hash)]
+pub struct ParseLengthError {
```
If we're going to keep `from_*_slice` methods this should be called `FromSliceError` since the error has nothing to do with parsing text.

tcharding replied:
oh I didn't even notice, during refactoring to use `TryFrom` and remove `_slice` methods this error type is no longer used.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+        }
+
+        impl fmt::Display for $ty {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
```
The `Debug`/`Display` impls are a bit messy. One would expect that `Display` of `Uint256` display in decimal which it doesn't do. It'd be clearer to implement `LowerHex` instead (with only `0x` added if `f.alternate()`) and delegate to that here instead. Also would be nice to handle alignment and such.

tcharding replied:
Aight, final task after lunch!  Thanks for the review @Kixunil, very thorough and educational, I appreciate your efforts.

Kixunil reviewed (commented):
Yes, `Target` and `Work` make sense, the design is good. Just found some details.

tcharding commented on src/pow.rs:
````diff
@@ -0,0 +1,1314 @@
+            /// assert_eq!(a.to_le(), [0xdeadbeef, 0, 0, 0]);
+            /// ```
+            #[inline]
+            pub fn to_le(self) -> [u64; 4] { self.0.to_le() }
````
Removed `to_le` and `to_be`. You are implying here that `from_le` and `from_be` are ok, right? Was the old name `from_array` better, or perhaps `from_be_array` (same for 'le' obviously)?

Kixunil replied:
> You are implying here that `from_le` and `from_be` are ok

No I didn't mean to imply that. Those leak the representation too.

If we're going to have them there should be no preference: `[u8; 32]`, `[u16; 16]`, `[u32; 8]`, `[u64; 4]`, `[u128; 2]` should all work because having a preference leaks representation. (apart from bytes which are just for en/decoding)

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+            ) -> Result<usize, crate::io::Error> {
+                self.0.consensus_encode(w)
+            }
+        }
```
oh, you are correct, my bad.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1314 @@
+        impl From<u64> for $ty {
+            #[inline]
+            fn from(x: u64) -> Self { $ty(Uint256::from(x)) }
+        }
```
Done, including unit tests.

tcharding commented:
I didn't all your suggestions implemented today @Kixunil, I've pushed what I've got done so as any new reviewers will see the most up to date code.

Still to do:

- Sort out debug/display
- Implement `Mul<u64>` and `Mul<u128>` for `Target` and `Work`
- Decide whether or not internal little-endian representation is a good idea
- Resolve overflow in `Target::difficulty()`

apoelstra commented:
Nice! Agree with the suggestion to switch to an internal BE encoding. Otherwise no strong opinions.

apoelstra commented:
@tcharding you also need to remove the u128_fuzz test from `fuzz/Cargo.toml`

tcharding commented:
> Implement `Mul<u64>` and `Mul<u128>` for `Target` and `Work`

I've done `Mul<$int>` where int is u8, u16, u32, u64. Have not implemented `Mul<u128>`

tcharding commented:
> I'm pretty confident with all the new/improved unit tests that I can trivially switch it to big-endian,

I should be more careful when I speak - this took me most of yesterday to do :) Hats off to you @apoelstra for doing all the bit twiddling initially, you're a boss.

tcharding commented:
`concat!` doesn't seem to work with Rust 1.41 but its been in the language since Rust 1.0, anyone have any idea what I'm doing wrong?

ref: https://doc.rust-lang.org/std/macro.concat.html

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1687 @@
+            #[doc = concat!("Creates a [`", stringify!($ty), "`] from a big-endian array of bytes.")]
+            #[inline]
+            fn from(a: [u8; 32]) -> Self { $ty(Uint256::from(a)) }
+        }
```
Why did you write `From<{array}>` instead of inherent `from_{endianess} function? The endianess ambiguity footgun is pretty serious. If you want to avoid duplication:

```rust
pub trait FromArray<A>: sealed::FromArray where A: AsMut<[u8]> {
    fn from_be(array: A) -> Self;

    fn from_le(mut array: A) -> Self {
        array.as_mut().reverse();
        Self::from_be(array)
    }
}
```

And probably have inherent functions so that people don't have to bother importing the trait for something that silly.

apoelstra replied:
>The endianess ambiguity footgun is pretty serious.

This made me laugh, in a somewhat dark way .... Satoshi was not careful about this and the consequence 13 years later is that we're still fighting about why half the hashes in Bitcoin are reversed half the time. So the consequences could be pretty bad, indeed.

tcharding replied:
I've added extensive constuctors of form `from_be_array_u8` (u8, u16, u32, u64, u128 - le and be) and the same for the other direction (e.g. `to_be_array_u8`). Also added convenience methods `to/from_be_bytes` (u8) and `to/from_be` (u128). 

All roundtrip unit tested.

tcharding replied:
I wasn't able to decipher your code above @Kixunil, I've used macros instead.

tcharding replied:
I'm not enamored with the names, open to suggestions please.

Kixunil replied:
The trait was supposed to solve the naming - the inherent function could be named `from_le` and `from_be`.

Maybe more code helps:

`pub fn from_le<A>(array: A) -> Self where Self: FromArray<A>`

`impl `FromArray<[{unsigned}; N]> for Target { /* ... */ }`

tcharding replied:
Aah, this is the same trick we did in `LockTime::is_satisfied_by(t: T)`, amusingly this realisation came to me mid way through a show at the theater on the weekend :)

Kixunil reviewed (changes requested):
Ambiguous `From` is IMO a big footgun.

Regarding `concat!` AFAIK (old?) Rust doesn't support macros in attributes and since this looks like a parsing issue I believe we can't `#[cfg]` special comments for it. It seems the only workable solution is to say `Self` instead.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1889 @@
+//! Proof-of-work related integer types.
+//!
+//! Provides the [`Work`] and [`Target`] types that are use in proof-of-work calculations. The
+//! functions here are designed to be fast.
```
I'm not sure if I've broken this statement now with everything I've done?

Kixunil replied:
I think they are still fast but I'm not sure if that's a very important property.

apoelstra replied:
Agreed on both counts. I think "designed to be fast" is supposed to mean "will not cause a DoS vector if you are syncing a node and using these functions to vet potential header chains" ... but like, our arithmetic would have to be *really* slow for that to be your biggest problem.

Kixunil replied:
Saying it's designed to be safe to check headers would be fine, I think.

tcharding commented:
Even with all the macros I think the `Work` and `Target` types are quite readable, the `Uint256` type not so much :)

tcharding commented on src/blockdata/constants.rs:
```diff
@@ -55,10 +55,17 @@ pub const SUBSIDY_HALVING_INTERVAL: u32 = 210_000;
+/// possible target (`max_target()`). Remember highest target == lowest difficulty.
+///
+/// In Bitcoind this is ~(u256)0 >> 32 stored as a floating-point type so it gets truncated, hence
+/// the low 208 bits are all zero.
```
I'm not super good with floating point number representation but I think its because of the significand and the exponent, the low 208 bits get zero'ed.

tcharding replied:
The original comment ha, I believe, a typo so I went digging for why this number

> /// In Bitcoind this is insanely described as ~((u256)0 >> 32)

0 shifted to the right is still zero so this is u256::MAX, which is not what the code is? Or am I being braindead?

Kixunil replied:
Ah, so what confused me is that when decoding from float `u32` is normally used (bits field in `BlockHeader`), not `u256` I guess this is what previous comment refers to as "insane".

I guess I'd rather remove it from doc, at least public. We can have `//` comment explaining this better if we anticipate people comparing this to Core.

tcharding replied:
Ah yes, removed from public docs.

tcharding commented:
Changes in force push: Two minor docs improvements.

Kixunil commented:
One thing I don't like about the `pow` module name is it resembles exponentiation - `u32::pow` method and similar. But I don't know if this can be improved. `proof_of_work` is quite long. It's not a big issue but if someone has a good idea I'd be happy.

Also, I'd like to have a `CompactTarget` newtype too since it's used in `BlockHeader`. We probably don't need it to be super-large, just support conversions to/from `Target` (from probably should be something like `from_target_lossy`), *explicit* conversion between `u32`, and consensus (de)serialization.

It doesn't need to be in this PR.

apoelstra commented on src/pow.rs:
```diff
@@ -0,0 +1,1892 @@
+    };
+}
+
+/// A big-endian 256 bit integer representing work.
```
"big-endian" is an implementation detail and doesn't make sense as part of the doccomment. (Same for `Target` below)

tcharding replied:
Removed, thanks.

apoelstra commented on src/pow.rs:
```diff
@@ -0,0 +1,1892 @@
+    const ONE: Uint256 = Uint256([0, 1]);
+
+    /// Creates `Self` from a big-endian array of `u128`s.
+    fn from_be_array_u128(a: [u128; 2]) -> Uint256 { Uint256(a) }
```
This is cool but why do we have a gazillion different conversion functions for different integer types? Why not just bytes? (And even then it's not totally obvious what the purpose is...serialization I guess).

apoelstra commented on src/pow.rs:
```diff
@@ -0,0 +1,1892 @@
+
+            /// Creates `Self` from a big-endian array of `u128`s.
+            #[inline]
+            pub fn from_be(array: [u128; 2]) -> $ty { $ty::from_be_array_u128(array) }
```
Related to my other comment -- why does the name `from_be` go to the `u128` conversion? Internal storage is an implementation detail, and intuitively I would expect a 256-bit type to be convertible to/from bytes rather than any intermediate-sized integer.

tcharding commented:
> Also, I'd like to have a `CompactTarget` newtype too since it's used in `BlockHeader`.

What do you mean "since its used in `BlockHeader`"? No fields in `BlockHeader` use are of type `Target`.

tcharding commented:
> One thing I don't like about the `pow` module name is it resembles exponentiation - `u32::pow` method and similar. But I don't know if this can be improved. `proof_of_work` is quite long. It's not a big issue but if someone has a good idea I'd be happy.

Duly noted, I have no better name ATM.

tcharding commented:
Loads of conversion code removed. Please note addition of new method `Target::is_met_by`.

Kixunil commented:
>  No fields in `BlockHeader` use are of type `Target`

The `bits` field should be `CompactTarget`, infallibly convertible to `Target`.

Kixunil commented:
How about calling the module `work`? `work::Work` is stuttering but not uncommon in Rust, `work::Target` may be slightly strange but not too bad. There isn't anything about *proving* work anyway, just verifying. So more logically `vow`, but that's incomprehensible.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1540 @@
+
+        let ret = !*self / self.wrapping_inc();
+        ret.wrapping_inc()
+    }
```
When making this const fn we can't trivially use the array trick because of possible division by zero. A workaround:

```rust
let mapped = [*self, Self::ZERO][self.is_max() as usize];
```

then work with `mapped` instead of self and conditionally return the appropriate value at the end.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1540 @@
+                    ret[idx as usize] += word >> (128 - bit_shift);
+                }
+            }
+        }
```
Now that we have two `u128` this can be simplified:

```rust
let left = self.0[0].wrapping_shl(shift) | self.0[1].wrapping_shr(128 - shift);
let right = self.0[1].wrapping_shl(shift);
Uint256([left, right])
```

Similarly below.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1540 @@
+
+/// Big-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+struct Uint256(pub [u128; 2]);
```
For consideration: use `struct U256(u128, u128);` instead:

* naming consistent with other types (minus casing)
* now that there are only two items we don't need for loops that much and seems even simpler without them
* We avoid awkward `self.0[0]` and `self.0[1]` - just have `self.0` and `self.1`

tcharding replied:
I like this idea, will play with it.

Kixunil commented on src/pow.rs:
```diff
@@ -0,0 +1,1540 @@
+            shift -= 1;
+        }
+
+        (Uint256(ret), sub_copy)
```
Oh, crap now that I wrote all the ideas how to make it `const fn` I realized I don't see how to fix this. Maybe my math skills are not good enough or maybe it's impossible.

tcharding replied:
I had a go at const'ing loads of things before. I don't think its worth the added complexity (and effort) at the moment. Rust 1.41 is too painful for widespread const'ing of functions in my opinion.

Kixunil reviewed (commented):
Would like constructors to be `const fn` if possible. If division (and thus `Target` <-> `Work` conversions) can't be const we can leave them for now and add conditional support in a subsequent PR conditioned on Rust 1.46.

We can use [a macro](https://github.com/Kixunil/hex_lit/blob/fad6e885b8bb4e80fd7c29fb24f1c6d8f173bb82/src/lib.rs#L77-L91) to conditionally make fns const (will need to be modified to support attributes)

Kixunil commented:
Forgot to say that apart from possible module renaming, I consider the code acceptable - it's already a strict improvement and the API looks good/unlikely to break. My suggestions can be done in subsequent PRs.

tcharding commented on src/blockdata/constants.rs:
```diff
@@ -55,10 +55,16 @@ pub const SUBSIDY_HALVING_INTERVAL: u32 = 210_000;
+/// ref: <https://en.bitcoin.it/wiki/Target>
+// In Bitcoind this is ~(u256)0 >> 32 stored as a floating-point type so it gets truncated, hence
+// the low 208 bits are all zero.
+pub fn max_target() -> Target { Target::from(0xFFFF_u64) << 208_u32 }
```
I did not do that because I thought using `Target::MAX` is confusing if one is thinking of the target as an integer type because it would be easy to assume it was the value with all bits set like other stdlib integer types.

apoelstra replied:
I don't think it'd be confusing

Kixunil replied:
We could add the invariant that `Target` is never above `MAX` and then it'd be natural but then alost all constructors would have to start returning `Result` or truncate the value. Maybe we should call it `CONSENSUS_MAX` (and avoid having `MAX` to avoid problems).

tcharding replied:
Breaking the "only push once a day" rule; I moved the function out of `contstants` and onto `Target`, and renamed it `max`. Making this a `const` is non-trivial for reasons discussed below (basically Rust 1.41.1).

Kixunil replied:
> non-trivial

Hmm? `Target(U256(0xFFFF_u128 << (208 - 128), 0))` Is there any problem with this?

tcharding commented on src/pow.rs:
````diff
@@ -0,0 +1,1540 @@
+    /// let w = Work::from(0xdeadbeef_u32);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(self) -> Target { Target(self.0.inverse()) }
````
I gave up on const functions because Rust 1.41 is really restrictive e.g. no `if` statements, no `for` statements, and even for functions that can be const I figured making them so really restricts how we can modify the functions internally. Thoughts?

tcharding replied:
Since we have not really started using const functions throughout the codebase it feels like something that can be done as a single block of work (PRs or epic) as opposed to now during this PR?

Kixunil replied:
> making them so really restricts how we can modify the functions internally. Thoughts?

What it mainly restricts is doing syscalls, which we will never do in integer types. There are other temporary restrictions depending on Rust version, the most painful being lack of helpful APIs.

As you may have noticed I mainly wanted constructors on the grounds that it allows usage of the type in const contexts. Unfortunately it spilled due to some constructors using math.

While I do agree with the general idea of adding `const` later, I'd prefer to have reasonably simple constructors `const` from the beginning to avoid the overhead of several PRs. So for this PR, I'd like `from_le_bytes` and `from_be_bytes` to be `const`. It's enough to use the types in const context and people can calculate manually and feed the bytes if needed. While writing the cont code is a bit annoying, I did it for you, so you just need to copy it and review. :)

tcharding replied:
I just saw this comment, will try to work this in to next version.

tcharding commented:
> Forgot to say that apart from possible module renaming, I consider the code acceptable - it's already a strict improvement and the API looks good/unlikely to break. My suggestions can be done in subsequent PRs.

Thanks man, I might have a play with the `U256(u128, u128)` form, it seems like a nice improvement and easier to do now while I'm fresh with the code.

tcharding commented:
> How about calling the module `work`? `work::Work` is stuttering but not uncommon in Rust, `work::Target` may be slightly strange but not too bad. There isn't anything about _proving_ work anyway, just verifying. So more logically `vow`, but that's incomprehensible.

I rekon the module name `pow` is meaningful for a top level module in a bitcoin repository, it doesn't really clash with 'power' IMO.

tcharding commented:
Implemented review suggestions:

- Now includes a `CompactTarget` type with the usual `to/from_consensus` methods and public conversion methods by way of `From` and `Into` for `Target` <-> `CompactTarget` conversions. I left the conversion logic in the now private methods on `Target`.
- Re-named `Uint256` -> `U256`

tcharding commented:
Thanks for all the suggestions @Kixunil, worthy improvements indeed.

tcharding commented:
Changes in force push: Move `constants::max_target()` to `Target::max()`. No other changes.

tcharding commented:
Force push: remove uint128 from `fuzz.yml` and do minor improvement to `Target::max` docs.

apoelstra reviewed (approved):
utACK 2929deeea6ef04f150f625ed372acbc6e8b8c4aa

Kixunil reviewed (commented):
Oh, crap I had this review in progress yesterday and forgot to submit...

tcharding commented:
Rebase only, no other changes.

sanket1729 reviewed (commented):
Did a somewhat detailed review checking for overflows, casts, and panics 60% of the PR. Mostly looks good. Will ACK tomorrow after reviewing the remaining 

Can we move the #cfg(test) code to the mod tests or to the very bottom?

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+    pub fn is_met_by(&self, hash: BlockHash) -> bool {
+        use crate::hashes::Hash;
+        let hash = Target(U256::from_le_bytes(hash.into_inner()));
+        hash <= *self
```
Oh yes that's much nicer, thanks.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+        let mut cur = *self;
+
+        loop {
+            let digit = (cur.rem(TEN)).low_u128() as u8; // Cast after rem 10 is lossless.
```
Can't remember, probably while I was playing with trying to const things (because we can't const trait methods AFAICT). Will change to `%`.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+impl_mul_for_uint256!(u8);
+impl_mul_for_uint256!(u16);
+impl_mul_for_uint256!(u32);
+impl_mul_for_uint256!(u64);
```
Because we multiply half a 'word' at a time using stdlib integer types. mul by `u32` uses a `u64`, mul by `u64` uses a `u128` etc., to implement mul by `u128` I'd have to work out another method of multiplying - which I was too lazy to do.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+    };
+}
+impl_hex!(LowerHex, "{:032x}");
+impl_hex!(UpperHex, "{:032X}");
```
Oh that was on purpose, I thought that for `Target` and `Work` we always wanted to see all the bits. I guess if reading the `U256` code alone it would seem strange but really this is just implemented on `U256` to save code duplication. Could improved by doing either
1. Have the macro take a type and implement for `Target` and `Work` directly - remove implementation for `U256` or add a 'correct' implementation
2. Just document why its implemented like this and leave it at that

Which would you prefer?

Kixunil replied:
Actually implementing it correctly is pretty straightforward. Just write the hex to temporary `[u8; 64]` buffer, convert to str using `from_utf8` and call `f.pad_integral()`. I'm preparing some hex PRs to make things easier so maybe wait for them.

tcharding replied:
Cool, I'll leave it as is and improve along with other hex improvements to come.

tcharding commented:
> Did a somewhat detailed review checking for overflows, casts, and panics 60% of the PR. Mostly looks good. Will ACK tomorrow after reviewing the remaining

Thanks man!
 
> Can we move the #cfg(test) code to the mod tests or to the very bottom?

Oh sure, that's a good idea.

tcharding commented:
> Hmm? Target(U256(0xFFFF_u128 << (208 - 128), 0)) Is there any problem with this?

Ha! so simple in hindsight, thanks. Added `Target::MAX` and also `max_value` for Rust 1.41.1 compatibility.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+    #[cfg(test)]
+    fn bit_at(&self, index: usize) -> bool {
+        if index > 255 {
+            return false;
```
Done!

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+
+        let d = Target::max().0 / self.0;
+
+        if d > MAX {
```
Done as suggested, is cleaner. Thanks.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Compact target is a 32-bit encoding of a 256 bit target.
```
Sure thing, done.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+        }
+
+        compact | (size << 24) as u32
+    }
```
Oh yes, I learned that during all the `to_u32` work and did not apply it here. Thanks.

tcharding replied:
I made `to_compact` and `from_compact` both public, and use `CompactTarget` instead of `u32`. This means users now have available the following:
```rust
// let bits = <get these off the wire>;
let t = Target::from_compact(CompactTarget::from_consensus(bits));
let t: Target = CompactTarget::from_consensus(bits).into();
```

tcharding replied:
In general is this idea of providing multiple ways of doing a thing something we want to support or should we stick to being opinionated and providing only a singe way (like we do for `to_consensus`). Or is it a case by case decision?

Kixunil replied:
It is generally useful to have at least `to_`/`into_` methods equivalent to the impl `Into<T>` for commonly used conversions to help inference.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+    /// Creates a [`U256`] from a little-endian array of `u8`s.
+    fn from_le_bytes(a: [u8; 32]) -> U256 {
+        let little = u128::from_le_bytes(a[..16].try_into().expect("16 bytes"));
+        let big = u128::from_le_bytes(a[16..].try_into().expect("slice is valid len"));
```
Nice, I added a function `split_in_half`.

tcharding commented on src/pow.rs:
```diff
@@ -0,0 +1,1510 @@
+    ($ty:ident) => {
+        impl From<$ty> for U256 {
+            fn from(x: $ty) -> U256 {
+                let little = u128::from(x);
```
No idea but it does not seem to :)

tcharding commented:
A bunch of changes in force push implementing suggestions from review above by @Kixunil and @sanket1729 as well as rebase to fix trivial merge conflict.

tcharding commented:
Please note the change to public API to now include `Target::to_compact` and `Target::from_compact`.

apoelstra reviewed (dismissed):
ACK b4473e97fa9e74bff569421751d0df6bafae9a0c

tcharding commented:
Rebase and fix merge conflicts only, no other changes.

tcharding commented:
And fix trailing whitespace :(

apoelstra reviewed (dismissed):
ACK 699dbe4a8091cfd08d0a8df6facc7e3e2bd3c2ea

tcharding commented:
Rebased and fixed some more pesky trailing whitespace issues - its hard to get good help :)

tcharding commented:
And ran formatter

apoelstra reviewed (approved):
ACK 863f6844a65852c0ba027b3da9880fbd7fd77d51

sanket1729 reviewed (approved):
code review ACK 863f6844a65852c0ba027b3da9880fbd7fd77d51. This looks great. Thanks for all the tests. Increases the confidence in the code. 

Would like to wait for @Kixunil's ACK to merge this because he was the most involved in this PR.

tcharding commented:
No rush, cheers.

tcharding commented:
Rebase only, no other changes.

apoelstra reviewed (dismissed):
ACK ada01c956a5ef10a61e06f3162b5db288394a674

Kixunil commented on bitcoin/src/blockdata/block.rs:
```diff
@@ -68,90 +66,32 @@ impl BlockHeader {
     }
 
     /// Computes the target [0, T] that a blockhash must land in to be valid.
```
OT: This looks like an array, got me confused.

tcharding replied:
Good point, fixed.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+        impl_from_unsigned_type!($ty, u16);
+        impl_from_unsigned_type!($ty, u32);
+        impl_from_unsigned_type!($ty, u64);
+        impl_from_unsigned_type!($ty, u128);
```
Oh, this will make it possible to call `42.into()` to get `Work` or `Target`. I was mostly against allowing that. I suggest an inherent method `from_raw<T: Into<u128>>(value: T)`. But still I'm not even sure if it's ever useful to work with such small numbers.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+    /// The maximum possible target (see [`Target::MAX`]).
+    ///
+    /// This is provided for compatibility with Rust 1.41.1, newer code should use [`Target::MAX`].
+    pub fn max_value() -> Self { Target::MAX }
```
Rust 1.41.1 doesn't disallow using just `MAX`, so this isn't needed.

tcharding replied:
Oh I thought it would be easier on devs writing code targeting 1.41.1 if they didn't have to use `max_value` sometimes and `MAX` sometimes. The comment "compatibility" is probably too strong. Can remove though if deemed unnecessary.

Kixunil replied:
Maybe say "consistency". (I only got the idea now, sorry.)

tcharding replied:
Good idea, used as suggested.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+
+        // The mantissa is signed but may not be negative.
+        if mant > 0x7F_FFFF {
+            Target::default()
```
I find this confusing. What is a default target? We should have a meaningful constant, not the trait.

tcharding replied:
I added private const `Target::ZERO`.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
```
How is dividing `Target` by itself meaningful? Looks like this should be removed.

tcharding replied:
Removed, thanks.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+            type Output = Self;
+            #[inline]
+            fn shr(self, shift: u32) -> Self { Self(self.0 >> shift) }
+        }
```
Bit operations on things that we no longer consider bags of bits are quite surprising. I'd rather remove them.

tcharding replied:
Fair point, removed. Cheers.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
```
Shouldn't these panic with overflow checks enabled? I suggest to use `debug_assertions` for now and maybe we could do this insane trick later: https://users.rust-lang.org/t/detecting-overflow-checks/67698/8

tcharding replied:
Oh I didn't know that overflow paniced in debug buids, TIL. I'll have to add back in `checked_add/sub/mul` to the `U256` type.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+/// Encoding of 256-bit target as 32-bit float.
+///
+/// This is used to encode a target into the block header. Satoshi made this part of consensus code
+/// in the original version of Bitcoin, likely copying an idea from OpenSSL.
```
Not necessarily for now but I'd like to see a more detailed explanation at some point.

apoelstra replied:
OpenSSL's bignum (`BN`) type has an encoding, which is even called "compact" as in bitcoin, which is exactly this format.

Kixunil replied:
I know what it is, I meant to expand the docs.

tcharding replied:
Add @apoelstra's comment to the code verbatim.

Kixunil replied:
Sorry for being unclear, I meant to explain roughly what the format is.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+}
+
+impl From<Target> for CompactTarget {
+    fn from(t: Target) -> Self { t.to_compact() }
```
One potentially surprising property of this is that `x == Target::from(CompactTarget::from(x)` doesn't always hold.

Maybe we should call it `to_compact_lossy` and skip this `From` impl. Maybe also implement `TryFrom` returning error if any information would be lost (but that' seems too much).

tcharding replied:
Good point, I've done as suggested and further commented `to_compact_lossy` also.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+
+/// Big-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+struct U256(u128, u128);
```
A non-doc comment saying field `0` is high, `1` is low could be useful for outsiders.

tcharding replied:
Added `// (high, low): u.0 contains the high bits, u.1 contains the low bits.`

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+        use crate::hashes::hex::ToHex;
+        let bytes = self.to_be_bytes();
+        if serializer.is_human_readable() {
+            serializer.serialize_str(&bytes.to_hex())
```
TODO: fast hex encoding.

tcharding replied:
Added todo comment.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+                let v = crate::prelude::Vec::from_hex(s)
+                    .map_err(|_| de::Error::invalid_value(de::Unexpected::Str(s), &self))?;
+
+                let mut b = [0_u8; 32];
```
`FromHex<[u8; 32]>` is available, so you could use that instead and remove the vast majority of this function along with an allocation. :)

tcharding replied:
Done!

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1530 @@
+            deserializer.deserialize_str(Visitor)
+        } else {
+            deserializer.deserialize_bytes(Visitor)
+        }
```
I think overloading a single visitor like this is unusual and they probably should be separate. If you don't want to create a new type maybe just carry a flag inside to error if unexpected one is used.

apoelstra replied:
Lol, I went through the codebase expecting to disagree with this, but actually the only case I could find of reusing a visitor like this is in the old `uint` code.

tcharding replied:
Fixed, I copied the code in `util::key`.

Kixunil reviewed (commented):
Unfortunately, despite the code being very good I still found some issue. Many of them nits but some more serious.

apoelstra commented:
Agreed with all of @Kixunil's comments, though I don't feel super strongly about any of them.

Kixunil commented:
To express myself better I've added emojis:

* :confused: - I'm not very sure about this but seems preferred to do it
* :eyes: - I consider this footgun and should be addressed
* no emoji - nit/OT/can be done later

tcharding commented:
Thanks for the review @Kixunil, thorough as ever. I'll re-work the PR. You the man.

tcharding commented:
Re-worked with suggestions from @Kixunil, a couple things to note:
1. This still does not have const constructors, I was unable to work out how to do that.
2. There are two ugly macros I can't get rid of, what I wanted to write was
```rust
impl<T: Into<u64>> Mul<U256> for T {
    type Output = U256;
    fn mul(self, rhs: T) -> Self {
        let (res, overflow) = rhs.mul_u64(self.into());
        debug_assert!(!overflow, "U256 multiplied by integer overflowed");
        res
    }
}
```
But I get:
```
error[E0210]: type parameter `T` must be covered by another type when it appears before the first local type (`U256`)
   --> bitcoin/src/pow.rs:691:6
    |
691 | impl<T: Into<u64>> Mul<U256> for T {
    |      ^ type parameter `T` must be covered by another type when it appears before the first local type (`U256`)
    |
    = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local, and no uncovered type parameters appear before that first local type
    = note: in this case, 'before' refers to the following order: `impl<..> ForeignTrait<T1, ..., Tn> for T0`, where `T0` is the first and `Tn` is the last
```

Kixunil commented:
1. OK, I don't want to hold this for too long, so I will try myself in a subsequent PR (probably with a bunch of others)
2. Oh, actually it's impossible to do that due to coherence rules. You could try `trait UintMarker: Into<u64> {} impl UintMarker for u8, u16, u32, u64 {}` and then use `UnitMarker` bound, but macros are OK too.

tcharding commented:
Changes in force push:
- Removed `FIXME`s
- Fixed stale macro name integer identifier (`uint256` -> `u256`)
- Used `Div<Into<u128>>` instead of into `u64`

tcharding commented:
Used "consistency" instead of "compatibility" when documenting `max_value()`.

tcharding commented:
@Kixunil, I tried the `UintMarker` idead but could not get it to work. Left the macros in there for now.

I believe all review suggestions have been seen to and this one is ready to go. 

One thing that may be missing, we are not providing any way to parse a string (hex or otherwise). With the current flux in the hex stuff I'm inclined to leave it as is and add that in later?

Kixunil commented:
> With the current flux in the hex stuff I'm inclined to leave it as is and add that in later?

Yeah, that's probably the best way.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1537 @@
+                type Value = U256;
+
+                fn expecting(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                    f.write_str("a bytestring")
```
Nit: "a sequence of bytes" would sound better.

tcharding replied:
Ok, that's three times in the last week I've balked at something, second guessed myself, and then its been picked up in review - I guess that means I'm learning :) Will fix since I have to rebase to fix conflicts anyways.

Kixunil reviewed (approved):
ACK e6913389590e583eb00979830067321943da3c6b

apoelstra reviewed (dismissed):
ACK e6913389590e583eb00979830067321943da3c6b

tcharding commented:
Changes in force push:
- Rebase, fix conflicts in CI fuzz.yml
- Use "a sequence of bytes" as suggested
- Improve the two Mul macros using `$(,)?` (as done in #1297)

apoelstra reviewed (approved):
ACK 01c79d1e69460c440c7aab605a85616777bfad63

apoelstra reviewed (approved):
ACK cb9893c4a974e3bed8288871d8e93d5aab087c32

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1536 @@
+/// a block. (See also [`Work`].)
+///
+/// ref: <https://en.bitcoin.it/wiki/Target>
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
```
Why is `Default` still here?

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1536 @@
+/// Work is a measure of how difficult it is to find a hash below a given [`Target`].
+///
+/// ref: <https://en.bitcoin.it/wiki/Work>
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
```
`Work` being zero seems less bad than `Target` but I'm still unsure.

Kixunil reviewed (commented):
This is frustrating but I really don't think we should release `Target` with `Default` impl. It'd be nice to remove it in a separate commit.

apoelstra commented:
If we add another commit which just removes `Default` I'll happily ACK it, it will take 20 seconds out of my day.

tcharding commented:
Done, I've no clue ATM if the `Default` was discussed before but if so and I omitted removing it previously then my apologies, either way, its gone now! Cheers

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -130,14 +130,14 @@ impl Sub for Work {
-    // When parsing nBits, Bitcoin Core converts a negative target threshold into a target of zero.
-    const ZERO: Target = Target(U256::ZERO);
+    /// When parsing nBits, Bitcoin Core converts a negative target threshold into a target of zero.
+    pub const ZERO: Target = Target(U256::ZERO);
```
Shouldn't this be `pub(crate)` instead? Zero target seems very strange to use by consumers and we only use it in tests.

apoelstra replied:
Since it's used as a clamp in consensus code it may be useful for users who want to compare against it, or who want to test their own block-handling code in pathological cases.

It's clearly commented as being a degenerate/edge case. I think it's ok.

tcharding replied:
Yeah users wanting to compare to zero was what made me make it public.

apoelstra reviewed (approved):
ACK 02a2b43b2b0729aa7b18afbdc90be8654a959790

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -0,0 +1,1536 @@
+
+        impl fmt::Display for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
```
This formats `Work` and `Target` as decimals. Did we really intend that?

tcharding replied:
I can't remember what made me do this but it was intentional at the time because I remember writing the `fmt_decimal` function. Whether or not we want to do this I don't know right at the moment (my brain is in holiday mode I don't really know whats going on, just checking my PRs to push trivial things forward).

Kixunil replied:
OK, I guess let's go with it. They are numbers, not hashes so formatting them as hex is not obvious and there's no better way to explicitly format as decimals.

apoelstra replied:
I also don't remember this discussion, but my guess is that because Bitcoin Core logs these values as decimals, we decided to go with that

Kixunil replied:
Oh, that's great to hear we're not the only strange ones doing that. :)

Also :tada: for merging this!

apoelstra replied:
I think Core also logs the log2 value as a float, which is an interesting idea but probably needs its own discussion and bikeshedding if we wanted to do it in `Display`. (Maybe it should be the alternate display? I dunno.)

Kixunil replied:
Maybe just provide conversion to `f64`?

apoelstra replied:
I like that (called `.log2()` or something, not `.into` ;)).

Kixunil replied:
`log2` sounds good.

Kixunil reviewed (approved):
ACK 02a2b43b2b0729aa7b18afbdc90be8654a959790

However, I'd like a confirmation that `Display` formatting as decimal is intentional.

sanket1729 commented:
@tcharding, special callout for sticking through with this PR over multiple months
